dds/DCPS/TypeAssignability.cpp

```diff
@@ -0,0 +1,2083 @@
+      MemberId id_a = ta.struct_type.member_seq.members[i].common.member_id;
+      const NameHash& h_a = ta.struct_type.member_seq.members[i].detail.name_hash;
+      ACE_CDR::ULong
+        name_a = (h_a[0] << 24) | (h_a[1] << 16) | (h_a[2] << 8) | (h_a[3]);
+      for (size_t j = 0; j < tb.struct_type.member_seq.members.size(); ++j) {
+        MemberId id_b = tb.struct_type.member_seq.members[j].common.member_id;
+        const NameHash& h_b = tb.struct_type.member_seq.members[j].detail.name_hash;
+        ACE_CDR::ULong
+          name_b = (h_b[0] << 24) | (h_b[1] << 16) | (h_b[2] << 8) | (h_b[3]);
+        if ((name_a == name_b && id_a != id_b) ||
+            (id_a == id_b && name_a != name_b)) {
+          return false;
+        } else if (name_a == name_b && id_a == id_b) {
+          matched_members.push_back(std::pair<const MinimalStructMember*,
+                                    const MinimalStructMember*>
+                                    (&ta.struct_type.member_seq.members[i],
+                                     &tb.struct_type.member_seq.members[j]));
+          break;
+        }
+      }
+    }
+
+    // There is at least one member m1 of T1 and one corresponding member
+    // m2 of T2 such that m1.id == m2.id
+    if (matched_members.size() == 0) {
+      return false;
+    }
+
+    // For any member m2 of T2, if there is a member m1 of T1 with the same
+    // ID, then the type KeyErased(m1.type) is-assignable-from the type
+    // KeyErased(m2.type)
+    for (size_t i = 0; i < matched_members.size(); ++i) {
+      const CommonStructMember& member = matched_members[i].second->common;
+      const MinimalTypeObject* toa = 0;
+      const MinimalTypeObject* tob = 0;
+      bool type_matched = false;
+      // KeyErased(T) is defined only for struct and union
+      if (get_struct_member(&tob, member)) {
+        if (!get_struct_member(&toa, matched_members[i].first->common)) {
+          return false;
+        }
+        type_matched = true;
+      } else if (get_union_member(&tob, member)) {
+        if (!get_union_member(&toa, matched_members[i].first->common)) {
+          return false;
+        }
+        type_matched = true;
+      }
+
+      if (type_matched) {
+        MinimalTypeObject key_erased_a = *toa, key_erased_b = *tob;
+        erase_key(key_erased_a);
+        erase_key(key_erased_b);
+        if (!assignable(key_erased_a, key_erased_b)) {
+          return false;
+        }
+      }
+    }
+
+    // Members for which both optional is false and must_understand is true in
+    // either T1 or T2 appear in both T1 and T2.
+    // Members marked as key in either T1 or T2 appear in both T1 and T2.
+    for (size_t i = 0; i < ta.struct_type.member_seq.members.size(); ++i) {
+      MemberFlag flags = ta.struct_type.member_seq.members[i].common.member_flags;
+      MemberId id = ta.struct_type.member_seq.members[i].common.member_id;
+      bool found = false;
+      if ((flags & IS_OPTIONAL) == 0 &&
```

Comment:
The two sides of the `&&` could be combined into one mask/compare operation.  But I'm OK letting the optimizer do that if it wants to.  I'm not sure one way would be easier to read and maintain than the other.

Reply:
I've changed to combine them into one comparison. It's a bit harder to read but I think it is ok with a comment.